[REVIEW] Histogram prim

teju85 opened the pull request:
Histogram primitive with various implementations underneath:
1. global mem atomics based
2. shared mem atomics based (with different bit-sizes per bin)
3. shared mem atomics with `__match_any_sync` intrinsic
4. shared mem hash based implementation

Credits should duly go to Lars Nyland and @canonizer

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,488 @@
+  auto op = [=] __device__(int binId, IdxT idx) {
+    atomicAdd(sbins + binId, 1);
+  };
+  histCoreOp<DataT, BinnerOp, IdxT, VecLen>(data, n, binner, op);
```
Processing multiple elements per thread is required to get the benefit from shared memory atomics.

The simplest approach is to always launch a fixed number of thread blocks, e.g. `2 * n_multiprocesors`, and distribute the data among them.

This also applies to other kernels.

teju85 replied:
this should be fixed in the commits I'll be pushing soon. Thanks for pointing it out.

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,488 @@
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+void smemHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
+              int tpb, cudaStream_t stream) {
+  int nblks = ceildiv<int>(VecLen ? n / VecLen : n, tpb);
```
It's better to use a fixed number of thread blocks (see the previous comment).

teju85 replied:
done.

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,488 @@
+  auto iword = binId / WORD_BITS;
+  auto sh = binId % WORD_BITS;
+  auto old_word = atomicXor(sbins + iword, unsigned(1 << sh));
+  if (old_word >> sh & 1) atomicAdd(&bins[binId], 2);
```
`if ((old_word >> sh & 1) != 0)`

teju85 replied:
Done.

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,488 @@
+                   IdxT n, int tpb, cudaStream_t stream, BinnerOp op) {
+  size_t bytes = n * sizeof(DataT);
+  if (n <= 0) return;
+  if (16 / sizeof(DataT) && bytes % 16 == 0) {
```
`if (sizeof(DataT) <= 16 && bytes % 16 == 0)`

The same applies to other comparisons here.

teju85 replied:
good catch. The correct check here should have been:
`if (16 % sizeof(DataT) == 0 && bytes % 16 == 0) {`
Have fixed this one. Thanks

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,488 @@
+    histogramVecLen<DataT, BinnerOp, IdxT, 8 / sizeof(DataT)>(
+      type, bins, nbins, data, n, tpb, stream, op);
+  } else if (4 / sizeof(DataT) && bytes % 4 == 0) {
+    histogramVecLen<DataT, BinnerOp, IdxT, 4 / sizeof(DataT)>(
```
Will it work if `sizeof(DataT) == 3`, but `n % 4 == 0`, and thus `bytes % 4 == 0`?

teju85 replied:
Most use-cases will be either 32b/64b int's or float/double. For all these cases, this check shouldn't have a problem. But you're right, better safe than to be sorry. See my previous response for the fix. Thanks.

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,488 @@
+template <typename IdxT>
+HistType selectBestHistAlgo(IdxT nbins) {
+  size_t smem = maxSharedMem();
+  size_t requiredSize = nbins * sizeof(unsigned);
```
This whole function should definitely be simplified.

teju85 replied:
if you have any suggestions, I'm all ears.

canonizer replied:
At the very least, the cascade of ifs can be replaced by a loop, with the return value being selected from an array. It may also be possible to compute the index into that array directly.

teju85 replied:
Done.

teju85 commented:
@canonizer This is now ready for another round of your review.

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,468 @@
+
+template <typename IdxT, int VecLen>
+dim3 computeGridDim(IdxT nrows, IdxT ncols) {
+  const auto maxBlks = 2 * getMultiProcessorCount();  // assume occupancy of 2
```
For just 256 threads per block, feel free to have more blocks, especially for small numbers of columns and bins.

This can also be done in a different PR.

teju85 replied:
I have generalized this through the use of cudart's occupancy API.

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,468 @@
+  int nCeil = alignTo<int>(nrows, stride);
+  typedef TxN_t<DataT, VecLen> VecType;
+  VecType a;
+  a.fill(DataT(0));
```
Do we need to fill it with zeros?

teju85 replied:
Thanks. This was a debug artifact. Removed it.

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,468 @@
+
+template <unsigned BIN_BITS>
+DI void incrementBin(unsigned* sbins, int* bins, int nbins, int binId,
+                     int col) {
```
Is it possible to drop the `col` parameter, and pass `bins + col * nbins` as `bins` instead?

The same applies to `incrementBin<1>`.

teju85 replied:
The catch is for computing `iword` you'd need `binId`, and when the smem-bin overflows, we'd need `col` and `binId` to know where to update the overflowed bin. If we just pass `bins`, we'd have to use modulo operation to know the `binId`. Thus, explicitly passing `col` as an argument.

canonizer replied:
My point is to replace calls to `incrementBin<Bits::BIN_BITS>(sbins, bins, (int)nbins, binId, col)` with `incrementBin<Bits::BIN_BITS>(sbins, bins + col * nbins, (int)nbins, binId)`, and to replace `bins + binOffset` within `incrementBin()` with just `bins`. `binId` would remain the same.

teju85 replied:
Done.

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -0,0 +1,468 @@
+__global__ void smemBitsHistKernel(int* bins, const DataT* data, IdxT nrows,
+                                   IdxT nbins, BinnerOp binner) {
+  extern __shared__ unsigned sbins[];
+  constexpr unsigned WORD_BITS = sizeof(unsigned) * 8;
```
These constants get repeated, so feel free to wrap them into a type parameterized by `BIN_BITS`. This is up to you, however.

teju85 replied:
Done.

canonizer commented on cpp/src_prims/utils.h:
```diff
@@ -230,6 +230,15 @@ IntType gcd(IntType a, IntType b) {
+  int devId;
+  CUDA_CHECK(cudaGetDevice(&devId));
+  cudaDeviceProp prop;
+  CUDA_CHECK(cudaGetDeviceProperties(&prop, devId));
```
Isn't it possible to use `cudaDeviceGetAttribute()` instead?

teju85 replied:
Ah.. this function is not used anywhere. We use `getSharedMemPerBlock` instead. Let me remove this one.

canonizer commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -410,25 +424,12 @@ HistType selectBestHistAlgo(IdxT nbins) {
-  if (requiredSize <= smem) {
-    return HistTypeSmemBits1;
+  for (int bits = 16; bits >= 1; bits >>= 1) {
+    auto nBytesForBins = bits * nbins / 8;
```
`ceildiv(bits * nbins, 8)`

teju85 replied:
Done.

teju85 commented:
@canonizer PR is now ready for another round of review.

teju85 commented:
Tagging @lnyland-nvidia to help us reviewing, since he wrote the initial version of hash-based smem hist algo.

@lnyland-nvidia the hash-based version starts from Line 271 onwards in the file `cpp/src_prims/stats/histogram.h`. I had to modify your initial version to also add support for doing histograms across multiple columns (like in a columnar dataset).

lnyland-nvidia commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -67,8 +65,39 @@ void gmemHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
+  }
+  __syncthreads();
+  for (auto i = tid; i < nbins; i += blockDim.x) {
+    atomicAdd(bins + i, sbins[i]);
```
Do you want to check if the sbins[i] is > 0?  Could save significant global atomics.

teju85 replied:
Thanks. I missed this one. Fixed now.

lnyland-nvidia commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -67,8 +65,39 @@ void gmemHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
+                               BinnerOp binner) {
+  extern __shared__ int sbins[];
+  auto tid = threadIdx.x + IdxT(blockIdx.x) * blockDim.x;
+  for (auto i = tid; i < nbins; i += blockDim.x) {
```
Where is nbins defined?

teju85 replied:
Yes, this was an oversight and got fixed in the later commits.

lnyland-nvidia commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -236,6 +236,27 @@ void histogramImpl(HistType type, int* bins, IdxT nbins, const DataT* data,
+  }
+  if (nbins % 4096 == 0) {
+    return HistTypeGmemSwizzle;
+  }
```
You can use swizzle on any size, you just have to adjust the size to the next multiple of 4096.  Something like nbins = ((nbins + 4096)>>12) <<12;

teju85 replied:
Hi Lars, since we also need to swizzle the histogram bins for this approach, I kept this algo for future optimization work. Will certainly make a note of this when we start working on this swizzle-based implementation. Thank you.

lnyland-nvidia commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -175,21 +175,92 @@ void smemBitsHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
+    if (count >= hashSize) {
+      idx = -1;
+      break;
+    }
```
This is a sequential search for an empty slot in the HT, right?  And if every entry is looked at and none is found, then the HT is flushed (in the calling code).  I have a feeling this will be very slow, but if you have tests that show it isn't, then please leave it be.  In the worst case, 1024 threads will fill the HT in 5 iterations (5*1024 < 6047).  Perhaps a better solution is to have count be less than some small value, like 30 or 100, so if you have 30 attempts to find a bin and you don't succeed, you flush.

teju85 replied:
Valid point! This count threshold needs to be set to properly balance between the penalty of finding an empty and the cost of flushing the hash table. Let me do that.

teju85 replied:
In the interest of time, I have reduced this threshold to a low value of 10. (meaning 10 tries from a thread, before it declares "we need a flush"). We certainly need to experimentally decide this number.

lnyland-nvidia commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -175,21 +175,92 @@ void smemBitsHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
+    if (needFlush[0]) {
+      flushHashTable(ht, hashSize, bins);
+      if (threadIdx.x == 0) {
+        needFlush[0] = 0;
```
This is too soon.  You could have a warp that hasn't read needFlush[0] before it is cleared, right?  The assignment to needFlush[0] may happen immediately, not at the syncthreads call.

teju85 replied:
The `flushHashTable` is guarded by a syncthreads-sandwich for this purpose. Maybe I should just remove the syncthreads call there and put it here for avoiding this confusion for others. Let me do that.

lnyland-nvidia commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -175,21 +175,92 @@ void smemBitsHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
+    }
+    if (iNeedFlush) {
+      int hidx = findEntry(ht, hashSize, binId);
+      atomicAdd(&(ht[hidx].y), 1);
```
Are you guaranteed to get a valid index here?  I think you can make a solid argument that after the flush, there are 6047 bins available, and that there should be enough room for every thread with iNeedFlush to get a bin.  I'll bet it is true 99.999999999% of the time.  Imagine a warp not getting scheduled for a while and the other warps fill the hashtable before this warp is scheduled.  Couldn't findEntry() return -1?

lnyland-nvidia replied:
Or maybe not.  If the loop goes around again, each thread in all the other warps will find a bin, increment it, and hit the syncthreads() call.  At that point, the thread with iNeedFlush set will have to make progress, so at most 1024-32 bins will have been used.  I think you're ok.

teju85 replied:
Yup agreed. We don't need another check for valid index here. Thanks.

teju85 replied:
I have left a comment at this place to help remind ourselves of this in the future.

lnyland-nvidia commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -175,21 +175,92 @@ void smemBitsHist(int* bins, IdxT nbins, const DataT* data, IdxT n, BinnerOp op,
-  auto maxSmem = maxSharedMem();
-  auto maxHashEntries = int(maxSmem / sizeof(int2));
+  // NOTE: assumes 48kB smem!
+  int hashSize = 6047;
```
You could make this work for a variety of shmem sizes with a cascading if structure.  Something like `if (maxSmem > 96k) hashSize = (some prime number bigger than 96k/8) else if (maxSmem > 64k) hashSize = (some prime number up to 96k) else if (maxSmem > 48k) ... else hashSize = 993;

lnyland-nvidia replied:
This is ugly.  I'm sure you can come up with a pretty solution based on a handful of smem sizes.

teju85 replied:
Fair enough. It should be easy to do that automatically. Let me do it.

teju85 replied:
Added a simple seive-based technique to automatically figure out the best hash table size for a given shared memory configuration.

lnyland-nvidia replied:
I didn't mean to suggest that you write an entire sieve.  I was thinking of something more like this:
auto maxSmem = maxSharedMem();
auto maxHashEntries = int(maxSmem / sizeof(int2));
int hashSize = 997;
if (maxHashEntries >= 16381) hashSize = 16381;
if (maxHashEntries >= 32749) hashSize = 32749;
if (maxHashEntries >= 49139) hashSize = 49139;
if (maxHashEntries >= 65521) hashSize = 65521;
if (maxHashEntries >= 98299) hashSize = 98299;
if (maxHashEntries >= 131071) hashSize = 131071;
if (maxHashEntries >= 196657) hashSize = 196657;

lnyland-nvidia replied:
Except I used the wrong values here.  I was thinking in bytes, not hash entries.

teju85 replied:
I see! We'll need such a thing in future at other places too. So, since this is already written, I'll just keep this one. :) Thank you for catching this one.

teju85 commented:
@lnyland-nvidia I have addressed most of your issues, except for the right threshold to decide when to flush. I'll finish it sometime today/tomorrow. In the meantime, please re-review the rest.

teju85 commented:
rerun tests

teju85 commented:
@lnyland-nvidia I have, for now, reduced this flush threshold to a small number of 10, in the interest of getting this PR done sooner. We certainly need to run experiments to figure out the right threshold param. Can you re-review now please and let me know if this looks good to be merged?

lnyland-nvidia commented on cpp/src_prims/stats/histogram.h:
```diff
@@ -331,7 +329,9 @@ __global__ void smemHashHistKernel(int* bins, const DataT* data, IdxT nrows,
     }
     __syncthreads();
     if (needFlush[0]) {
+      __syncthreads();
```
Do you need this sync?  You've just done one to get needFlush[0] shared among all threads.  What does the middle one do?  I actually thought it was correct without these here nor needing the ones in flushHashTable.

teju85 replied:
My bad. You're right. This syncthreads is not needed. Removed it. Thanks.

teju85 commented:
rerun tests

teju85 commented:
@lnyland-nvidia are you ok with merging this to cuML?

lnyland-nvidia commented:
@teju85 Yes, please merge.

teju85 commented:
Thank you Lars.

@dantegd please merge.
